### ISSUE TITLE
Incorrect handling of lateout pairs in inline asm

### ISSUE THREAD

**Issue by newpavlov** [rust-lang/rust#101346]
(issue is loosely owned, in terms of P-high tracking, by @wesleywiser and @pnkfelix keeping their eyes on https://github.com/llvm/llvm-project/issues/57550)

### Updated bug description
The following Rust function:
```rust
pub fn foo() -> u32 {
    let t1: u32;
    let t2: u32;
    unsafe {
        asm!(
            "mov {0:e}, 1",
            "mov eax, 42",
            lateout(reg) t1,
            lateout("eax") t2,
            options(nostack),
        );
    }
    t1
}
```
Gets compiled into this obviously incorrect assembly:
```asm
example::foo:
        mov     eax, 1
        mov     eax, 42
        ret
```
Godbolt link: https://rust.godbolt.org/z/Yb9v7WobM

LLVM incorrectly reuses register for a pair of `lateout`s if it can see that one of those does not get used later.

### Original description below

We get spurious segfaults in the `chacha20` crate when we run tests compiled for i686 target (`i686-unknown-linux-gnu` to be exact), see https://github.com/RustCrypto/stream-ciphers/issues/304 for more information. Interestingly enough, Rust 1.56 does not have this issue, only 1.57 and later. [Changes](https://github.com/RustCrypto/utils/pull/792) in the `cpufeatures` crate which revealed the issue look quite innocent. The issue also disappears if `zeroize` feature is disabled, which is quite tangential to `cpufeatures` (it only adds `Drop` impls). Even weirder, @aumetra's [findings](https://github.com/RustCrypto/stream-ciphers/issues/304#issuecomment-1235837221) show that segfault happens at the following line:
```rust
let mut buf = [0u8; MAX_SEEK];
```

Granted, the `chacha20` crate contains a fair bit of unsafe code, as well as its dependencies, so the issue may be caused by unsoundness somewhere in our crates. But the circumstantial evidence makes a potential compiler bug quite probable.

**Comment by talchas** [comment]
Yeah, the 32-bit cpuid implementation at <https://doc.rust-lang.org/src/core/up/up/stdarch/crates/core_arch/src/x86/cpuid.rs.html#62> is compiling to something wrong. This is compiling to 
```asm
   0x565ba631 <+193>:   mov    %ebx,%eax
   0x565ba633 <+195>:   cpuid  
   0x565ba635 <+197>:   xchg   %eax,%ebx
```
on `rustc 1.63.0 (4b91a6ea7 2022-08-08)`
which of course does actually clobber %ebx, because apparently a `lateout` register can overlap an explicitly-chosen `inlateout` register. I don't know if that's an intentional LLVM thing, an LLVM bug, or what. According to godbolt this asm!() results in
```
%0 = tail call { i32, i32, i32, i32 } asm sideeffect "movl %ebx, ${0:k}\0Acpuid\0Axchgl %ebx, ${0:k}", "=r,={ax},={cx},={dx},1,2,~{memory}"(i32 1, i32 1) #1, !dbg !10, !srcloc !16
```

**Comment by newpavlov** [comment]
Yeah, after your post in the RustCrypto issue I also tried it in godbolt and got the same weird assembly: https://rust.godbolt.org/z/4z9nrY1Eh

Rust:
```rust
use std::arch::x86::__cpuid;

pub unsafe fn foo() -> u32 {
    let res = __cpuid(0);
    res.ebx
}
```
Generated assembly which is obviously wrong:
```asm
example::foo:
        xor     eax, eax ; set eax to zero
        xor     ecx, ecx ; set eax to zero
        mov     eax, ebx ; "cache" calee-saved ebx to eax
        cpuid ; WHOOPS not only we now use eax which contains ebx's value,
                  ; but also CPUID overwrites eax, making the caching useless
        xchg    eax, ebx ; "restore" cached ebx
        ret
```
Bad news is that on x68-64 we [get](https://rust.godbolt.org/z/fY5YhxzEe) the same wrong assembly:
```asm
example::foo:
        xor     eax, eax
        xor     ecx, ecx
        mov     rax, rbx
        cpuid
        xchg    rax, rbx
        ret
```

**Comment by talchas** [comment]
It's supposed to pick a different register and save it there, and that both should work (with proper clobbers) and should be better than pushing to the stack. Changing it to `out(reg) ebx,` works in godbolt, though I can't swear that it is guaranteed as opposed to just avoiding this example. `out(reg) ebx, lateout("eax") eax,` does still work, so it presumably still won't be able to conflict with the non-input `edx`. I'm still not sure if this is an LLVM bug or intentionally part of the rules of inline asm (constraint rules have never been very clear).

**Comment by talchas** [comment]
<https://rust.godbolt.org/z/Gr1ve77a6> suggests that it might be that LLVM considers _unused_ lateout arguments to be fair game to delete along with the clobber.

**Comment by newpavlov** [comment]
I think the issue is with incorrect use of `lateout(reg) ebx,` together with `inlateout("eax") leaf => eax,`. Because of the `late` specifiers LLVM assumes that it's legal to use for both these registers (i.e. it assigns `{0}` to `eax`).

**Comment by talchas** [comment]
Yeah, the thing is that that doesn't really make sense in general - `lateout(reg) ebx, lateout("eax") eax` (which still fails) really shouldn't be able to assign them to be the same register any more than it could for `lateout(reg) x, lateout(reg) y`.

**Comment by newpavlov** [comment]
Indeed, LLVM does something weird with `lateout` registers when one of them does not get used.

https://rust.godbolt.org/z/3e6MzzWfa

Rust:
```rust
pub fn foo() -> u32 {
    let eax: u32;
    let ebx: u32;
    unsafe {
        asm!(
            "mov {0}, 1",
            "mov {1}, 42",
            lateout(reg) eax,
            lateout(reg) ebx,
        );
    }

    eax
}
```

Assembly:
```asm
example::foo:
        push    rax

        mov     rax, 1
        mov     rcx, 42

        pop     rcx ; ??????
        ret
```

**Comment by talchas** [comment]
That one is probably just `pop arbitrary_register` - it needs to have push/poped _something_ around the asm for stack alignment reasons since it isn't marked nostack.

**Comment by newpavlov** [comment]
You are right. A better example, which probably can be used as minified demonstration of the issue: https://rust.godbolt.org/z/dfP4s9dG7
```rust
pub fn foo() -> u32 {
    let t1: u32;
    let t2: u32;
    unsafe {
        asm!(
            "mov {0}, 1",
            "mov eax, 42",
            lateout(reg) t1,
            lateout("eax") t2,
            options(nostack),
        );
    }
    t1
}
```
```asm
example::foo:
        mov     rax, 1
        mov     eax, 42
        ret
```

**Comment by newpavlov** [comment]
Either way, the `late` specifiers do not make much sense in the case of CPUID intrinsics and I think they should be simply removed. We only need one scratch register to cache EBX, which is not EAX, EBX, ECX, or EDX.

**Comment by comex** [comment]
For what it's worth, the original `asm!` block seems to me like it would misbehave if `{0}` happened to be assigned as `ebx` itself.  Why does it do this backup and restore rather than just specifying `ebx` as a clobber?

But that's not what's happening here, so it does seem like LLVM is miscompiling the code.

**Comment by talchas** [comment]
You can't specify `ebx` as a constraint because LLVM uses it internally for something (maybe exactly this plt cache? I forget) and doesn't handle clobbers of it like a sane compiler (iirc you get compile errors sometimes).

**Comment by talchas** [comment]
That said, I'm not sure that that's true anymore - while I vaguely recall some sort of error along those lines, I can't trigger one from the PLT usage of `ebx`. <https://github.com/rust-lang/rust/issues/90083> doesn't give details and I can't find them quickly re LLVM/clang.

**Comment by newpavlov** [comment]
Trying to clobber `rbx` in the usual way currently results in the following compilation error:
```
error: cannot use register `bx`: rbx is used internally by LLVM and cannot be used as an operand for inline asm
```
Honestly, it's quite annoying and I wish we did not have such exception.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23101346.20Potential.20miscompilation.20on.20i686.20of.20chacha20)).

@rustbot label -I-prioritize +P-critical T-compiler

**Comment by pnkfelix** [comment]
Does the change that landed in rust-lang/stdarch#1329 resolve this? (In other words, do we just need to pull in those changes in some manner to resolve this issue?)

Or is there something else that will need to happen?

**Comment by pnkfelix** [comment]
also, given that this is a P-critical issue that affects 1.57 and later: Should we be talking about beta backports of the changes associated with fixing this?

**Comment by talchas** [comment]
That change should resolve this, yes.

**Comment by Amanieu** [comment]
The issue affects all uses of `asm!` where `lateout` operands are used. As far as I can tell `cpuid` is the only one that is affected in the standard library, but uses of `asm!` in other crates could also be affected by this bug.

**Comment by apiraino** [comment]
Issue was visited during T-compiler meeting on Zulip ([notes](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-09-22/near/300177066)). Demoting priority to `P-high` since #101861 reduced the impact and the LLVM upstream fix seems to be formalized

@rustbot label -P-critical P-high

**Comment by pnkfelix** [comment]
T-compiler P-high review. Nikita volunteered to be default assignee for LLVM miscompilations.

@rustbot assign @nikic

**Comment by newpavlov** [comment]
Looks like this issue was fixed in LLVM 19 (Rust 1.82+).
